elixir/lib/cloner.ex

```diff
@@ -0,0 +1,36 @@
+defmodule GithubRepoCloner.Cloner do
+  @http_client Application.compile_env(:github_repo_cloner, :http_client)
+  @system Application.compile_env(:github_repo_cloner, :system)
+
+  def clone(nil), do: true
+
+  def clone(username) do
+    username
+    |> request_repo_info
+    |> parse_response
+    |> clone_repos
+  end
+
+  def clone, do: true
+
+  defp request_repo_info(username) do
+    @http_client.get("https://api.github.com/users/#{username}/repos")
+  end
+
+  defp parse_response({:ok, %Tesla.Env{status: 200, body: body}}) do
+    elem(Poison.decode(body), 1)
+  end
+
+  defp parse_response({:ok, %Tesla.Env{status: _}}), do: []
+
+  defp clone_repos(repo_info) do
+    repo_info
+    |> Enum.map(fn(item) -> "git clone #{item["clone_url"]}" end)
+    |> Enum.join(" & ")
+    |> run_command
+  end
+
+  defp run_command(command) do
+    @system.cmd("sh", ["-c", command])
```

Comment:
We have one complete command as an argument here, but `System.cmd` expects an executable defined in PATH as the first argument, followed by command arguments: https://hexdocs.pm/elixir/main/System.html#cmd/3

This is a workaround by using `sh` as the executable. Open to refactor ideas here.

Reply:
Interesting, @murjax 

## 1. performance (currently this looks like a bash solution)
I wonder how this competes with running each of the commands in a separate elixir processes.
😅 `git clone #{item["clone_url"]} & git clone #{item["clone_url"]} &`

I have not tested yet but this might not hit the api limit due to a bash bottleneck instead of an elixir bottleneck.

This makes me think we might write a repo cloner written in bash that is 3 lines of code!
## 2. security
I would not extract this into a run command function.
Currently the input is controlled in that it gets the ["clone_url"] from the api request.
 
Making it a generic function invites a future developer to reach for this function for other solutions; and those solutions could not have a controlled input. While I cannot envision a use-case where this happens; I think it is always important to be aware security holes and try to avoid allowing future developers and users from making mistakes.

I think we can assume that the input is controlled you would need to be super nefarious to change the clone url to something like `https:yourrepo.git; rm -rf ~;`
... thinking about this further, this is logically a potential security hole.
We could close this security hole fix by removing all instances of `&`, `|`, or `;` from the clone url before building the sh command. Still build the command itself.

## Refactor suggestion:
I would spawn processes and run each git clone independently.
considerations:
1. Are we going to hit an IO bottleneck? (Q: which can make more network requests at one time elixir or bash?)
2. My guess is that a multi process elixir with any large number will be immediately banned by the api by going over the limit. I think the balance is to figure out what will cause a violation and be right at the edge.

Reply:
Q: If we switch to spawn will the tests still work?

Reply:
> Q: If we switch to spawn will the tests still work?

Current tests will fail if we switch to spawn, because `Cloner.clone/1` will no longer return anything.

> My guess is that a multi process elixir with any large number will be immediately banned by the api by going over the limit. 

I think the balance is to figure out what will cause a violation and be right at the edge.

> Are we going to hit an IO bottleneck? (Q: which can make more network requests at one time elixir or bash?)

These are all going to hit bash ultimately. We should write a spawn version and benchmark it.

We might be safe on requests coming from the `git` client: https://github.com/orgs/community/discussions/44515#discussioncomment-4795475. 
Only one API request is made to get the repo info.

> We could close this security hole fix by removing all instances of &, |, or ; from the clone url before building the sh command. Still build the command itself.

Good callout on security. Could we cleanse the url with a regex?